cloudconfig/sshinit/filetransport.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshinit
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"math/rand"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3/ssh"
+	"golang.org/x/sync/errgroup"
+)
+
+type FileTransporter struct {
+	params   ConfigureParams
+	prefix   string
+	safeHost string
+
+	errGroup *errgroup.Group
+	ctx      context.Context
+}
+
+// NewFileTransporter returns an SCP file transporter that implements a cloudinit.FileTransporter
+// to send payloads to the target machine, saving them in a temporary location.
+func NewFileTransporter(ctx context.Context, params ConfigureParams) *FileTransporter {
+	ft := &FileTransporter{
+		params: params,
+	}
+	eg, egCtx := errgroup.WithContext(ctx)
+	ft.errGroup = eg
+	ft.ctx = egCtx
```

Comment:
1. we shouldn't be doing this, 2. I can't see where ft.ctx is used anyway.

We should be deferring the construction of the errgroup to SendBytes - send bytes should take a context (not NewFileTransporter) and that ctx passed to SendBytes is passed to errgroup.WithContext(). The ctx in the result params can then be ignored because nothing in SendBytes uses a ctx

errGroup is only an attr on the struct so we can use it to call Wait(). We could instead remove the attr and have SendBytes return a wait func(). Then we don't need the Wait() on FileTransporter either.

Reply:
SendBytes is used by old cloudinit code that would need an overhaul to add in support for what you're describing. I'll remove the ctx from the transporter, but effectively this code is a 2 year old hack I wrote a while ago. The cloudinit code (which sometimes is not cloud init) already needs almost a full rewrite, that I won't be doing today.

Reply:
OK, I think I've made it better.